tools/codegen/api/autograd.py

```diff
@@ -42,6 +45,29 @@ class Derivative:
     # Saved outputs that are referenced by the formula.
     saved_outputs: Tuple[SavedAttribute, ...]
 
+# Represents a forward formula that calculates forward derivatives
+# for one tensor.
+@dataclass(frozen=True)
+class ForwardDerivative:
+    # The formula string (legit C++ expression).
+    # Note that special keywords such as "linear" or "element_wise" have been
+    # replaced by the automatically generated formula.
+    formula: str
+
+    # Name of the output argument for which this formula calculates forward
+    # derivatives
+    var_name: str
+
+    # Type of the output argument for which this formula calculates forward
+    # derivatives
+    var_type: Type
+
+    # Inputs for which the forward derivatives are required for this formula
+    required_inputs_fw_grad: Tuple[str, ...]
+
+    # Inputs for which the primal is required for this formula
+    required_inputs_primal: Optional[Tuple[str, ...]]
```

Comment:
How come this is Optional?

Reply:
Because some formulas don't need any of the primal value of any input. For example `add()` only needs the tangent from the input.